# 2025. 03. 25.

## Reactor 3 Reference Guide

### 테스팅

#### Reactor 디버깅

##### 3. 디버그 모드에서 스택 트레이스 읽기

`operatorStacktrace` 디버그 기능을 활성화해 처음 예시를 보면 다음과 같은 일이 일어난다:

1. 구독처와 나머지 덜 흥미로운 부분을 가리키는 스택 트레이스가 첫 프레임 이후로 잘려나간다.
2. 원본 예외에 특수한 억제된(suppressed) 예외가 추가된다(혹은 이미 있는 경우 바꾼다).
3. 메시지가 몇 가지 섹션과 함께 이 특수한 예외에 대해 만들어진다.
4. 첫 섹션은 실패한 연산자의 어셈블리 위치를 추적한다.
5. 두 번째 섹션은 이 연산자가 구축되고 오류가 전파된 체인을 보여주려고 시도한다.
6. 마지막 섹션은 원본 스택 트레이스이다.

출력된 전체 스택 트레이스는 다음과 같다:

```
java.lang.IndexOutOfBoundsException: Source emitted more than one item
    at reactor.core.publisher.MonoSingle$SingleSubscriber.onNext(MonoSingle.java:127) // 1. 원본 스택 트레이스는 한 프레임으로 잘린다.
    Suppressed: The stacktrace has been enhanced by Reactor, refer to additional information below: // 2. 새로운 부분: 스택을 포착한 래퍼 연산자를 볼 수 있다. 이곳이 트레이스백이 나타나기 시작한 곳이다.
Assembly trace from producer [reactor.core.publisher.MonoSingle] : // 3. 먼저 연산자가 어셈블된 곳에 대한 상세 정보를 얻는다.
    reactor.core.publisher.Flux.single(Flux.java:7915)
    reactor.guide.GuideTests.scatterAndGather(GuideTests.java:1017)
Error has been observed at the following site(s): // 4. 두 번째로 처음부터 마지막까지 오류가 전파된 연산자 체인에 대한 정보를 얻는다.
    *_______Flux.single ⇢ at reactor.guide.GuideTests.scatterAndGather(GuideTests.java:1017) // 5. 오류를 만난 각 연산자는 사용자 클래스와 사용된 줄 위치로 나타낸다. 이곳이 "루트"이다.
    |_ Mono.subscribeOn ⇢ at reactor.guide.GuideTests.debuggingActivated(GuideTests.java:1071) // 6. 체인의 일부이다.
Original Stack Trace: // 7. 스택 트레이스의 나머지 부분은 마지막으로 이동했다.
        at reactor.core.publisher.MonoSingle$SingleSubscriber.onNext(MonoSingle.java:127)
...
// 8. 연산자 내부의 일부를 보여준다(일부를 생략했다).
...
        at reactor.core.publisher.Mono.subscribeWith(Mono.java:4363)
        at reactor.core.publisher.Mono.subscribe(Mono.java:4223)
        at reactor.core.publisher.Mono.subscribe(Mono.java:4159)
        at reactor.core.publisher.Mono.subscribe(Mono.java:4131)
        at reactor.guide.GuideTests.debuggingActivated(GuideTests.java:1067)
```

포착된 스택 트레이스는 원본 오류 뒤에 억제된(supprressed) `OnAssemblyException`으로 추가된다. 여기에는 세 부분이 있지만 첫 번째 섹션이 가장 흥미롭다. 예외를 유발한 연산자에 대한 생성 경로를 보여준다. 여기서는 문제를 발생시킨 `single`이 실제로는 `scatterAndGather` 메서드에서 생성된 것임을 보여준다.

이제 원인을 찾기에 충분한 정보가 모였다. `scatterAndGather` 메서드를 보자:

```java
private Mono<String> scatterAndGather(Flux<String> urls) {
    return urls.flatMap(url -> doRequest(url))
           .single(); // 1. 이곳에 `single`이 있다.
}
```

이제 오류의 근본 원인이 몇몇 URL에 몇 번의 HTTP 호출을 수행하지만 너무 제한적인 `single`로 체이닝된 `flatMap`이라는 것을 알 수 있다. 잠깐의 `git blame` 후 코드 작성자와 논의한 끝에 원래 의도는 덜 제한적인 `take(1)`을 사용하는 것이었음을 알아냈다.

문제를 해결했다.

이제 스택 트레이스에서 다음 섹션을 보자:

```
Error has been observed at the following site(s):
```

오류가 (`subscribe`에 가장 가까운)체인의 마지막 연산자에서 발생했기 때문에 트레이스백의 두 번째 부분은 이 예시에서 그다지 흥미롭지는 않다. 다른 예시로 더 명확하게 보자:

```java
FakeRepository.findAllUserByName(Flux.just("pedro", "simon", "stephane"))
              .transform(FakeUtils1.applyFilters)
              .transform(FakeUtils2.enrichUser)
              .blockLast();
```

이제 `findAllUserByName` 안에 실패하는 `map`이 있다고 상상해보자. 여기서는 트레이스백의 두 번째 부분이 다음과 같을 것이다:

```
Error has been observed at the following site(s):
    *________Flux.map ⇢ at reactor.guide.FakeRepository.findAllUserByName(FakeRepository.java:27)
    |_       Flux.map ⇢ at reactor.guide.FakeRepository.findAllUserByName(FakeRepository.java:28)
    |_    Flux.filter ⇢ at reactor.guide.FakeUtils1.lambda$static$1(FakeUtils1.java:29)
    |_ Flux.transform ⇢ at reactor.guide.GuideDebuggingExtraTests.debuggingActivatedWithDeepTraceback(GuideDebuggingExtraTests.java:39)
    |_   Flux.elapsed ⇢ at reactor.guide.FakeUtils2.lambda$static$0(FakeUtils2.java:30)
    |_ Flux.transform ⇢ at reactor.guide.GuideDebuggingExtraTests.debuggingActivatedWithDeepTraceback(GuideDebuggingExtraTests.java:40)`
```

이는 오류를 고지받은 연산자 체인의 섹션에 해당한다:

1. 예외는 첫 번째 `map`에서 비롯됐다. 이는 `*` 연결자로 루트로 식별하고 `_`는 들여쓰기다.
2. 예외는 (둘 다 `findAllUserByName` 메서드에 해당하는)두 번째 `map`에서 관측됐다.
3. `filter`와 `transform`에서도 관측됐다. 체인의 일부가 재사용 가능한 변환(transformation) 함수(여기선 `applyFilters` 유틸리티 메서드)에 의해 만들어졌음을 나타낸다.
4. 마지막으로 `elapsed`와 `transform`에서 관측됐다. 한번 더 `elapsed`는 두 번째 변환의 변환 함수에 의해 적용됐다.

같은 예외가 여러 체인을 따라 전파되는 일부 사례에서는 "루트" 표시자 `*_`로 이러한 체인을 더 잘 식별할 수 있다. 여러 번 관측되면 호출처 정보에 `(observed * times)`가 될 것이다.

예를 들어, 다음 스니펫을 보자:

```
public class MyClass {
    public void myMethod() {
        Flux<String> source = Flux.error(sharedError);
        Flux<String> chain1 = source.map(String::toLowerCase).filter(s -> s.length() < 4);
        Flux<String> chain2 = source.filter(s -> s.length() > 5).distinct();

        Mono<Void> when = Mono.when(chain1, chain2);
    }
}
```

위 코드에서 오류는 별개의 두 체인 `chain1`과 `chain2`를 따라 `when`으로 전파된다. 트레이스백에는 다음이 포함될 것이다:

```
Error has been observed at the following site(s):
    *_____Flux.error ⇢ at myClass.myMethod(MyClass.java:3) (observed 2 times)
    |_      Flux.map ⇢ at myClass.myMethod(MyClass.java:4)
    |_   Flux.filter ⇢ at myClass.myMethod(MyClass.java:4)
    *_____Flux.error ⇢ at myClass.myMethod(MyClass.java:3) (observed 2 times)
    |_   Flux.filter ⇢ at myClass.myMethod(MyClass.java:5)
    |_ Flux.distinct ⇢ at myClass.myMethod(MyClass.java:5)
    *______Mono.when ⇢ at myClass.myMethod(MyClass.java:7)
```

여기서 알 수 있는 것은:

1. 세 개의 "루트" 요소가 있다(`when`이 진짜 루트이다).
2. `Flux.error`에서 시작한 두 체인을 볼 수 있다.
3. 두 체인 모두 같은 `Flux.error` 소스에 기반한 것으로 보인다(`observed 2 times`).
4. 첫 번째 체인은 `Flux.error().map().filter`이다.
5. 두 번째 체인은 `Flux.error().filter().distinct()`이다.

> [!TIP]
>
> **트레이스백과 억제된 예외에 대한 참고**: 트레이스백은 원본 오류 뒤에 억제된 예외로 추가되므로 이는 이 메커니즘을 사용하는 다른 예외 유형에 간섭할 수 있다. 이러한 예외는 `Exceptions.multiple(Throwable...)`을 통해 직접 만들거나 (`Flux#flatMapDelayError`처럼)여러 오류 소스를 조인하는 일부 연산자를 통해 생성할 수 있다. 트레이스백이 조합의 요소로 고려되고  반환된 `List`의 일부인 경우 이들은 `Exception.unwrapMultiple(Throwable)`을 통해 래핑되지 않은 `List`로 벗겨낼 수 있다. 어떤 식으로든 이것이 적절하지 않은 경우 `Exceptions.isTraceback(Throwable)` 검사로 트레이스백을 식별하고 `Exceptions.unwrapMultipleExcludingTracebacks(Throwable)`을 대신 사용해 벗겨낼 수 있다.

> [!WARNING]
>
> **정적 예외**: 예외 처리를 최적화하는 것은 때로 예외 상태가 발생할 때 새 인스턴스를 생성하는 대신 사전에 생성한 전역 정적 인스턴스를 재사용하는 것이 필요하기도 한다. 이렇게 재사용된 예외에서는 트레이스백에 추가된 억제된 예외가 누적되기 시작하므로 메모리 누수를 피할 수 없다. 대부분의 상황에서 정적 예외는 예외 생성의 가장 큰 비용 중 하나인 스택 트레이스 포착을 비활성화한다. 하지만 예외는 스택 트레이스 없이 약간의 메모리 할당 비용으로 동적으로 만들어질 수 있어 정적 인스턴스가 필요 없다. 여러분의 코드가 여전히 예외의 정적 인스턴스에 의존한다면 트레이스백에서 메모리 누수를 피하는 방법은 [억제를 비활성화하는 대신 이 예외가 리액티브 체인을 따라 전파됐을 때 트레이스백을 볼 수 없게 하는 것이다.

이 계측 방식과 스택 트레이스를 생성하는 것은 비용이 든다. 즉, 이 디버깅 기능은 통제된 방식으로 마지막 수단으로 사용해야 한다.



[javadoc-throwable]: https://docs.oracle.com/javase/8/docs/api/java/lang/Throwable.html#Throwable-java.lang.String-java.lang.Throwable-boolean-boolean-